src/controller/src/lib.rs

```diff
@@ -451,7 +451,8 @@ where
         + Codec64
         + Unpin
         + From<EpochMillis>
-        + TimestampManipulation,
+        + TimestampManipulation
+        + std::fmt::Display,
```

Comment:
Nit: `Timestamp` already derives `Debug`. Maybe we can use that and avoid the additional trait bound here?

Reply:
According to the compiler error that I was getting, this requirement comes from this function: https://github.com/MaterializeInc/materialize/blob/278bada8179ae91d5e9dfb5bdc09609c5c620870/src/storage-controller/src/lib.rs#L1904-L1918. Specifically, to convert `StorageError<T>` to `anyhow::Error`, `T` needs to implement `Display`. I can't figure out a way around this, any ideas?

Reply:
I think that comes from these two impls:

```
impl<T: Debug + Display + 'static> Error for StorageError<T> { ... }
impl<T: fmt::Display + 'static> fmt::Display for StorageError<T> { ... }
```

If you'd define both with only `T: Debug` bounds, that would avoid the need for `T: Display`. But then you couldn't use `Antichain::pretty`, unless you'd also change that to use the `Debug` impl instead. I don't think that's worth it.

But it is unusual that the `Error for StorageError<T>` needs both a `Debug` and a `Display` bound. I think you can just drop the former?

Reply:
> But it is unusual that the `Error for StorageError<T>` needs both a Debug and a Display bound. I think you can just drop the former?

I might be misunderstanding, but I think all errors need to implement both `Debug` and `Display`.

https://doc.rust-lang.org/core/error/trait.Error.html

> Errors must describe themselves through the [`Display`](https://doc.rust-lang.org/core/fmt/trait.Display.html) and [`Debug`](https://doc.rust-lang.org/core/fmt/trait.Debug.html) traits.

If I try and remove the bound then I get the following error:

```
error[E0277]: `T` doesn't implement `std::fmt::Debug`
   --> src/storage-types/src/controller.rs:210:38
    |
210 | impl<T: Display + 'static> Error for StorageError<T> {
    |                                      ^^^^^^^^^^^^^^^ `T` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
    |
note: required for `StorageError<T>` to implement `std::fmt::Debug`
   --> src/storage-types/src/controller.rs:164:10
    |
164 | #[derive(Debug)]
    |          ^^^^^ unsatisfied trait bound introduced in this `derive` macro
note: required by a bound in `StdError`
   --> /home/joe/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/error.rs:32:18
    |
32  | pub trait Error: Debug + Display {
    |                  ^^^^^ required by this bound in `Error`
    = note: this error originates in the derive macro `Debug` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
    |
210 | impl<T: Display + 'static + std::fmt::Debug> Error for StorageError<T> {
    |                           +++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `mz-storage-types` (lib) due to 1 previous error

```

Reply:
You are right, my bad!